packages/client/src/generation/dmmf.ts

```diff
@@ -108,96 +109,104 @@ class DMMFSchemaHelper implements Pick<DMMF.Document, 'schema'> {
     }
   }
 
-  resolveOutputTypes() {
-    for (const type of this.outputTypes.model) {
+  private resolveOutputTypes() {
+    this.resolveOutputTypesInNamespace('prisma')
+    this.resolveOutputTypesInNamespace('model')
+  }
+
+  private resolveOutputTypesInNamespace(namespace: DMMF.FieldNamespace) {
+    for (const type of this.outputTypes[namespace]) {
       for (const field of type.fields) {
-        if (typeof field.outputType.type === 'string' && !ScalarTypeTable[field.outputType.type]) {
-          field.outputType.type =
-            this.outputTypeMap[field.outputType.type] ||
-            this.outputTypeMap[field.outputType.type] ||
-            this.enumMap[field.outputType.type] ||
-            field.outputType.type
+        if (typeof field.outputType.type !== 'string') {
+          // should not happen
+          continue
         }
-      }
-      type.fieldMap = keyBy(type.fields, 'name')
-    }
 
-    for (const type of this.outputTypes.prisma) {
-      for (const field of type.fields) {
-        if (typeof field.outputType.type === 'string' && !ScalarTypeTable[field.outputType.type]) {
-          field.outputType.type =
-            this.outputTypeMap[field.outputType.type] ||
-            this.outputTypeMap[field.outputType.type] ||
-            this.enumMap[field.outputType.type] ||
-            field.outputType.type
+        if (field.outputType.location === 'scalar') {
+          continue
+        }
+        const fieldNamespace = field.outputType.namespace ?? 'model'
+        if (field.outputType.location === 'outputObjectTypes') {
+          field.outputType.type = this.outputTypeMap[fieldNamespace][field.outputType.type]
+        } else if (field.outputType.location === 'enumTypes') {
+          field.outputType.type = this.enumMap[fieldNamespace][field.outputType.type]
         }
       }
-
       type.fieldMap = keyBy(type.fields, 'name')
     }
   }
 
-  resolveInputTypes() {
-    const inputTypes = this.inputObjectTypes.prisma
-
-    if (this.inputObjectTypes.model) {
-      inputTypes.push(...this.inputObjectTypes.model)
-    }
+  private resolveInputTypes() {
+    this.resolveNamespaceInputTypes('model')
+    this.resolveNamespaceInputTypes('prisma')
+  }
 
-    for (const type of inputTypes) {
+  private resolveNamespaceInputTypes(namespace: DMMF.FieldNamespace) {
+    const types = this.inputObjectTypes[namespace] ?? []
+    for (const type of types) {
       for (const field of type.fields) {
         for (const fieldInputType of field.inputTypes) {
-          const fieldType = fieldInputType.type
-          if (
-            typeof fieldType === 'string' &&
-            !ScalarTypeTable[fieldType] &&
-            (this.inputTypeMap[fieldType] || this.enumMap[fieldType])
-          ) {
-            fieldInputType.type = this.inputTypeMap[fieldType] || this.enumMap[fieldType] || fieldType
+          if (typeof fieldInputType.type !== 'string') {
```

Comment:
Is this mostly for type narrowing?

Reply:
Yes, on engine side, `type` is always string but we define it is `string | SchemaEnum | SchemaObject` and override the value in one of the `resolveX` methods. Since we don't have separate types for input/output DMMF, TS types force us to handle non-string cases here as well.
Ideally, I'd just remove all `resolve` methods and leave original DMMF untouched (already created tech debt card for this). This was useful mostly for `query.ts`, for generator, "flat" types are more convenient for majority of cases.